contracts/treasury/src/handle.rs

```diff
@@ -125,15 +207,209 @@ pub fn try_register_asset<S: Storage, A: Api, Q: Querier>(
     })
 }
 
-pub fn rebalance<S: Storage, A: Api, Q: Querier>(
-    _deps: &mut Extern<S, A, Q>,
-    _env: &Env,
+pub fn register_allocation<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: &Env,
+    asset: HumanAddr,
+    alloc: Allocation,
 ) -> StdResult<HandleResponse> {
-    let messages = vec![];
+    let config = config_r(&deps.storage).load()?;
+
+    /* ADMIN ONLY */
+    if env.message.sender != config.admin {
+        return Err(StdError::Unauthorized { backtrace: None });
+    }
+
+    let full_asset = match assets_r(&deps.storage).may_load(asset.to_string().as_bytes())? {
+        Some(a) => a,
+        None => {
+            return Err(StdError::GenericErr {
+                msg: "Unregistered asset".to_string(),
+                backtrace: None,
+            });
```

Comment:
You can make these errors cleaner like this

```suggestion
            return Err(StdError::generic_err ("Unregistered asset"));
```